src/Microsoft.Framework.Runtime.Hosting/ProjectFilesCollection.cs

```diff
@@ -3,122 +3,103 @@
-            _preprocessSourcesMatcher.AddIncludePatterns(PreprocessPatterns);
-            _preprocessSourcesMatcher.AddExcludePatterns(SharedPatterns, ResourcesPatterns, ExcludePatterns);
-
-            _bundleExcludeFilesMatcher = new Matcher();
-            _bundleExcludeFilesMatcher.AddIncludePatterns(BundleExcludePatterns);
-
-            _resourcesMatcher = new Matcher();
-            _resourcesMatcher.AddIncludePatterns(ResourcesPatterns);
-
-            _sharedSourceMatcher = new Matcher();
-            _sharedSourceMatcher.AddIncludePatterns(SharedPatterns);
-
-            _contentFilesMatcher = new Matcher();
-            _contentFilesMatcher.AddIncludePatterns(ContentsPatterns);
-            _contentFilesMatcher.AddExcludePatterns(SharedPatterns,
-                                                    PreprocessPatterns,
-                                                    ResourcesPatterns,
-                                                    BundleExcludePatterns,
-                                                    SourcePatterns);
+            _projectFilePath = projectFilePath;
+
+            var excludeBuiltIns = PatternsCollectionHelper.GetPatternsCollection(rawProject, projectDirectory, projectFilePath, "excludeBuiltIn", DefaultBuiltInExcludePatterns);
+            var excludePatterns = PatternsCollectionHelper.GetPatternsCollection(rawProject, projectDirectory, projectFilePath, "exclude")
+                                                          .Concat(excludeBuiltIns);
+
+            var bundleExcludePatterns = PatternsCollectionHelper.GetPatternsCollection(rawProject, projectDirectory, projectFilePath, "bundleExclude", DefaultBundleExcludePatterns);
+
+            _sharedPatternsGroup = PatternsGroup.Build(rawProject, projectDirectory, projectFilePath, "shared", legacyName: null, fallback: DefaultSharedPatterns, buildInExcludePatterns: excludePatterns);
+
+            _resourcePatternsGroup = PatternsGroup.Build(rawProject, projectDirectory, projectFilePath, "resource", "resources", DefaultResourcesPatterns, excludePatterns);
+
+            _preprocessPatternsGroup = PatternsGroup.Build(rawProject, projectDirectory, projectFilePath, "preprocess", legacyName: null, fallback: DefaultPreprocessPatterns, buildInExcludePatterns: excludePatterns)
+                .ExcludeGroup(_sharedPatternsGroup)
+                .ExcludeGroup(_resourcePatternsGroup);
+
+            _compilePatternsGroup = PatternsGroup.Build(rawProject, projectDirectory, projectFilePath, "compile", "code", DefaultCompilePatterns, excludePatterns)
+                .ExcludeGroup(_sharedPatternsGroup)
+                .ExcludeGroup(_preprocessPatternsGroup)
+                .ExcludeGroup(_resourcePatternsGroup);
+
+            _contentPatternsGroup = PatternsGroup.Build(rawProject, projectDirectory, projectFilePath, "content", "files", DefaultContentsPatterns, excludePatterns.Concat(bundleExcludePatterns).Distinct())
```

Comment:
in case there are duplicate patterns. it can save some duplicate processing in matching, though the results won't be different.